## ETL hoja hechos

Realizo un análisis exploratorio de los datos, con la finalidad de encontrar patrones que permitan generar información para la toma de decisiones en relación a disminuir las víctimas fatales.

### Importo las librerias necesarias para comenzar el proceso de ETL 

In [1]:
import numpy as np
import pandas as pd
import herramientas
import warnings
warnings.filterwarnings("ignore")

Comienzo mi Analisis Exploratorio con la hoja **HECHOS** de mi base de datos 'homicidios'

In [2]:
df_hecho=pd.read_excel('data/homicidios.xlsx',sheet_name='HECHOS')
df_hecho

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                           AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   
..                                         ...           ...  ...     ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...     NaN   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...     NaN   
693                        AV. GAONA Y TERRADA       AVENIDA  ...     NaN   
694                         AV. EVA PERON 4071       AVENIDA  ...  4071.0   
695                    PADRE CARLOS MUJICA 709         CALLE  ...   709.0   

                                   Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada COMUNA  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
691                                  RIESTRA AV. y MOM      7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
693                                GAONA AV. y TERRADA     11   
694                                PERON, EVA AV. 4071      9   
695                            PADRE CARLOS MUGICA 709      1   

                                   XY (CABA)         pos x         pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
692    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
693   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
694    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
695  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      PARTICIPANTES    VICTIMA    A

In [3]:
df_hecho.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Altura', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')

Podemos observar que la hoja 'HECHOS' consta de 21 columnas y 696 filas inicialmente. A continuacion se puede observar el 'Diccionario de datos' con el cual podremos observar el significado de cada columna de nuestra base de datos

Diccionario de datos:
- **ID**:   identificador unico del siniestro
- **N_VICTIMAS**:	cantidad de víctimas
- **FECHA**:	fecha en formato dd/mm/aaaa
- **AAAA**:	año
- **MM**:	mes
- **DD**:	día del mes
- **HORA**:	hora del siniestro
- **HH**:	franja horaria entera
- **LUGAR_DEL_HECHO**:	Dirección del hecho
- **TIPO_DE_CALLE**:	Tipo de arteria. En el caso de intersecciones a nivel se clasifica según la de mayor jerarquía.

	- **calle**	Arteria cuya calzada tiene un ancho comprendido entre cinco (5) y trece (13) metros. Inlcluye pasajes.
	- **avenida**:	Arteria cuya calzada tiene un ancho total de por lo menos trece (13) metros.
	- **autopista**:	Vía multicarril con calzadas para ambas manos separadas físicamente, sin cruces a nivel, con accesos controlados y sin ingreso directo desde los predios frentistas lindantes. Incluye AU 25 de mayo, AU Perito Moreno, AU Dellepiane, Av Lugones, Av. Cantilo, AU Frondizi, AU Buenos Aires - La Plata en el tramo que circula dentro de la Ciudad de Buenos Aires y Paseo del Bajo. Incluye ingresos y egresos de las mismas y distribuidores.
	- **general paz**:	Avenida General Paz, ambos sentidos. Incluye ingresos y egresos de las mismas, distribuidores y calle colectora.

- **Calle**:	nombre de la arteria donde se produjo el hecho
- **Altura**:	altura de la arteria donde se produjo el hecho
- **Cruce**:	cruce en caso de que sea una encrucijada
- **Dirección Normalizada**:	direccion en formato normalizado USIG
- **COMUNA**:	Comuna de la ciudad (1 a 15)
- **XY (CABA)**:	geocodificación plana
- **pos x**:	longitud con separador punto. WGS84
- **pos y**:	latitud con separador punto. WGS84
- **PARTICIPANTES**:	conjunción de víctima y acusado
	- **MULTIPLE**:	cuando participan más de un vehículo como contraparte de la víctima.

- **VICTIMA**:	Vehículo que ocupaba quien haya fallecido a se haya lastimado a raíz del hecho, o bien peatón/a. Clasificación agregada del tipo de vehículos.

	- **PEATON**:	Víctima distinta de cualquier ocupante de un vehículo, ya sea un conductor/a o un pasajero/a. Se incluyen los ocupantes o personas que empujan o arrastran un coche de bebé o una silla de ruedas o cualquier otro vehículo sin motor de pequeñas dimensiones. Se incluyen también las personas que caminan empujando una bicicleta o un ciclomotor.
	- **MOTO**:	Vehículo a motor no carrozado que incluye motocicleta, ciclomotor y cuatriciclo.
	- **AUTO**:	Vehículo a motor destinado al transporte de personas, diferente de los motovehículos, y que tenga hasta nueve plazas (incluyendo al asiento del conductor) (Sedan, SUV, coupe, etc)
	- **CARGAS	Vehículo a motor destiando al transporte de cargas, incluye camiones pesados (con o sin acoplado o semirremolque, etc., camión de recolección de residuos) y livianos (utilitarios, furgonetas, pick-ups, camioneta con caja de carga).
	- **BICICLETA**:	Vehículo con al menos dos ruedas, que generalmente es accionado por el esfuerzo muscular de las personas que lo ocupan, en particular mediante pedales o manivelas. Incluye bicicletas de pedaleo asistido y/o con motor
	- **PASAJEROS**:	Personas lesionadas que se encuentran dentro, descendiendo o ascendiendo de las unidades de autotrasporte público de pasajeros/as y ómnibus de larga distancia
	- **MOVIL**:	Vehículos de emergencia: móviles policiales, ambulancias, autobombas. 
	- **OTRO**:	otros vehiculos

- **ACUSADO**:	Vehículo que ocupaba quien resultó acusado/a del hecho, sin implicar culpabilidad legal
	- **AUTO**:	Vehículo a motor destinado al transporte de personas, diferente de los motovehículos, y que tenga hasta nueve plazas (incluyendo al asiento del conductor) (Sedan, SUV, coupe, etc)
	- **BICICLETA**:	Vehículo con al menos dos ruedas, que generalmente es accionado por el esfuerzo muscular de las personas que lo ocupan, en particular mediante pedales o manivelas. Incluye bicicletas de pedaleo asistido y/o con motor
	- **CARGAS**:	Vehículo a motor destiando al transporte de cargas, incluye camiones pesados (con o sin acoplado o semirremolque, etc., camión de recolección de residuos) y livianos (utilitarios, furgonetas, pick-ups, camioneta con caja de carga).
	- **MOTO**:	Vehículo a motor no carrozado que incluye motocicleta, ciclomotor y cuatriciclo.
	- **OBJETO FIJO**:	Colisión contra objetos inmóviles fijados de manera permanente o semipermanente (columna, árbol, semáforo, etc.) o pérdidas de equilibrio de vehículos de dos ruedas que desencadenen la caída de sus ocupantes.
	- **PASAJEROS**:	Personas lesionadas que se encuentran dentro, descendiendo o ascendiendo de las unidades de autotrasporte público de pasajeros/as y ómnibus de larga distancia
	- **TREN**:	Equipo móvil que se desplaza exclusivamente sobre rieles
	- **OTRO**:	otros vehiculos
	- **SD**:	Sin datos sobre el vehículo participante



### A continuacion se procedera a realizar el proceso de ETL 

- **Estandarizacion del nombre de las columnas**: Dado que los nombres de las columnas algunos estan en minusculas, y el nombre de las mismas podria ser poco intuitivo a la hora de realizar analisis posteriormente, realizo la estandarizacion de los mismos

In [4]:
# Coloca la primera en mayúscula
df_hecho.columns = [x.capitalize() for x in df_hecho.columns]
# Reemplaza los guiones por espacios
df_hecho.columns = df_hecho.columns.str.replace('_', ' ')
# Renombra algunas columnas
df_hecho = df_hecho.rename(columns={'N victimas': 'Cantidad de victimas',
                                                      'Aaaa':'Año',
                                                      'Mm':'Mes',
                                                      'Dd':'Dia',
                                                      'Hh':'Hora en entero',
                                                      'Xy (caba)':'XY (CABA)',
                                                      'Victima': 'Victima'})


Verifico los cambios anteriores

In [5]:
df_hecho.columns

Index(['Id', 'Cantidad de victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Hora',
       'Hora en entero', 'Lugar del hecho', 'Tipo de calle', 'Calle', 'Altura',
       'Cruce', 'Dirección normalizada', 'Comuna', 'XY (CABA)', 'Pos x',
       'Pos y', 'Participantes', 'Victima', 'Acusado'],
      dtype='object')

Verifico la existencia de valores nulos y el tipo de dato 

In [6]:
herramientas.verifica_tipo_y_nulos(df_hecho)

nombre_campo                                         tipo_datos  \
0                      Id                                    [<class 'str'>]   
1    Cantidad de victimas                                    [<class 'int'>]   
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                     Año                                    [<class 'int'>]   
4                     Mes                                    [<class 'int'>]   
5                     Dia                                    [<class 'int'>]   
6                    Hora  [<class 'datetime.time'>, <class 'str'>, <clas...   
7          Hora en entero                     [<class 'int'>, <class 'str'>]   
8         Lugar del hecho                                    [<class 'str'>]   
9           Tipo de calle                                    [<class 'str'>]   
10                  Calle                   [<class 'str'>, <class 'float'>]   
11                 Altura                                  [<class 'float'>]   
12                  Cruce                   [<class 'str'>, <class 'float'>]   
13  Dirección normalizada                   [<class 'str'>, <class 'float'>]   
14                 Comuna                                    [<class 'int'>]   
15              XY (CABA)                                    [<class 'str'>]   
16                  Pos x                                    [<class 'str'>]   
17                  Pos y                                    [<class 'str'>]   
18          Participantes                                    [<class 'str'>]   
19                Victima                                    [<class 'str'>]   
20                Acusado                                    [<class 'str'>]   

    no_nulos_%  nulos_%  nulos  
0       100.00     0.00      0  
1       100.00     0.00      0  
2       100.00     0.00      0  
3       100.00     0.00      0  
4       100.00     0.00      0  
5       100.00     0.00      0  
6       100.00     0.00      0  
7       100.00     0.00      0  
8       100.00     0.00      0  
9       100.00     0.00      0  
10       99.86     0.14      1  
11       18.53    81.47    567  
12       75.43    24.57    171  
13       98.85     1.15      8  
14      100.00     0.00      0  
15      100.00     0.00      0  
16      100.00     0.00      0  
17      100.00     0.00      0  
18      100.00     0.00      0  
19      100.00     0.00      0  
20      100.00     0.00      0

  Se observa un total de 747 valores nulos en nuestro dataframe. El 81.47% de los valores de la columna 'Altura' son nulos, seguido del 24.57% de la columna 'Cruce', 1.15% de la columna 'Dirección normalizada' y 0.14% de la columna 'Calle'.

- **Columna: 'Altura'**

  Se considerara a la columna 'Altura' como dato no relevante debido a que para que se complete la misma, el 'hecho' debe ocurrir en algun punto de una calle y segun lo anterior , la mayoria de los hechos ocurren en las esquinas.

  Adicionalmente se puede observar que la informacion de la columna 'Altura', se encuentra de forma similar en las columnas 'Lugar del hecho'y 'Dirección. A continuacion dejo constancia del mismo.

In [7]:
df_hecho[['Lugar del hecho', 'Dirección normalizada','Altura']].head()

Lugar del hecho  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ   
1          AV GRAL PAZ  Y AV DE LOS CORRALES   
2                         AV ENTRE RIOS 2034   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA    

                               Dirección normalizada  Altura  
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...     NaN  
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.     NaN  
2                                ENTRE RIOS AV. 2034  2034.0  
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.     NaN  
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.     NaN

Procedo a eliminar la columna 'Altura'

In [8]:
df_hecho=df_hecho.drop('Altura', axis=1)
df_hecho.columns

Index(['Id', 'Cantidad de victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Hora',
       'Hora en entero', 'Lugar del hecho', 'Tipo de calle', 'Calle', 'Cruce',
       'Dirección normalizada', 'Comuna', 'XY (CABA)', 'Pos x', 'Pos y',
       'Participantes', 'Victima', 'Acusado'],
      dtype='object')

- **Columna: 'Cruce'**

Se observa que dicha columna, es categorica, la cual contiene 24.57% valores faltantes. La misma solo tiene datos cuando el hecho no ocurre en un cruce. Se relacionara esta columna junto con las de 'Lugar del hecho' y 'Tipo de calle', para observar los valores nulos de cruce en relación con 'Lugar del hecho' y 'Tipo de calle'

In [9]:
df_hecho[df_hecho['Cruce'].isnull()][['Lugar del hecho', 'Tipo de calle','Cruce']].head(10)

Lugar del hecho Tipo de calle Cruce
2                                  AV ENTRE RIOS 2034       AVENIDA   NaN
9                                  AV ENTRE RIOS 1366       AVENIDA   NaN
14                                       SUIPACHA 156         CALLE   NaN
33                                          LIMA 1483         CALLE   NaN
35  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....     AUTOPISTA   NaN
38                         AUTOPISTA LUGONES PK 10000     AUTOPISTA   NaN
42                            AV LUIS MARIA CAMPOS 30       AVENIDA   NaN
43                                          LIMA 1471         CALLE   NaN
50                           MAGARIÑOS CERVANTES 3900         CALLE   NaN
59                                   AV LA PLATA 2384       AVENIDA   NaN

Se observan los valores NO nulos de cruce en relación con 'Lugar del hecho' y 'Tipo de calle'

In [10]:
df_hecho[df_hecho['Cruce'].notnull()][['Lugar del hecho', 'Tipo de calle','Cruce']].head(10)

Lugar del hecho Tipo de calle  \
0           AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA   
1                   AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ   
3               AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA   
4           AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA   
5                      AV 27 DE FEBRERO Y AV ESCALADA       AVENIDA   
6                        NOGOYA Y JOAQUIN V. GONZALES         CALLE   
7                 AV GENERAL PAZ Y AV DE LOS CORRALES      GRAL PAZ   
8                  AV BELGRANO Y BERNARDO DE IRIGOYEN       AVENIDA   
10                         AV SCALABRINI ORTIZ Y VERA       AVENIDA   
11  AV REGIMIENTO DE PATRICIOS Y QUINQUELA MARTIN ...       AVENIDA   

                                  Cruce  
0   FERNANDEZ DE LA CRUZ, F., GRAL. AV.  
1                   DE LOS CORRALES AV.  
3              VILLEGAS, CONRADO, GRAL.  
4               SAENZ PE?A, LUIS, PRES.  
5                          ESCALADA AV.  
6                  GONZALEZ, JOAQUIN V.  
7                   DE LOS CORRALES AV.  
8                 IRIGOYEN, BERNARDO DE  
10                                 VERA  
11             QUINQUELA MARTIN, BENITO

Se observa que los datos faltantes efectivamente son los que se refieren a un hecho en algún punto de la calle que no es una esquina o intersección de dos calles. Por lo tanto, se decide modificar esta columna en un valor 'si' para los casos que son hechos en cruces de calles y 'no' en el caso que no sea un cruce de calles

In [11]:
df_hecho['Cruce'] = np.where(df_hecho['Cruce'].notnull(), 'Si', 'No')
# Se revia la columna
df_hecho[['Lugar del hecho', 'Tipo de calle','Cruce']]

Lugar del hecho Tipo de calle Cruce
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA    Si
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ    Si
2                           AV ENTRE RIOS 2034       AVENIDA    No
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA    Si
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA    Si
..                                         ...           ...   ...
691                          AV. RIESTRA Y MOM       AVENIDA    Si
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA    Si
693                        AV. GAONA Y TERRADA       AVENIDA    Si
694                         AV. EVA PERON 4071       AVENIDA    No
695                    PADRE CARLOS MUJICA 709         CALLE    No

[696 rows x 3 columns]

- **Columna: 'Datos Normalizados**

Dicha columna posee solo 8 datos nulos , por lo cual se decide modificar los valores NaN , SD (Sin Datos) debido a que no tenemos informacion de como se completo la misma. 

In [12]:
df_hecho['Dirección normalizada'].fillna("SD", inplace=True)
df_hecho[df_hecho['Dirección normalizada']== 'SD'][['Lugar del hecho', 'Tipo de calle','Dirección normalizada']]

Lugar del hecho Tipo de calle  \
38                AUTOPISTA LUGONES PK 10000     AUTOPISTA   
106         AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA   
119                                       SD         CALLE   
180  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     AUTOPISTA   
181                       AU DELLEPIANE 2400     AUTOPISTA   
313                 AUTOPISTA LUGONES KM 4.7     AUTOPISTA   
546             LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA   
621         AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA   

    Dirección normalizada  
38                     SD  
106                    SD  
119                    SD  
180                    SD  
181                    SD  
313                    SD  
546                    SD  
621                    SD

- **Columna: 'Id'**

Corroboro la existencia de valores duplicados 

In [13]:
herramientas.verifica_duplicados_por_columna(df_hecho, 'Id')

'No hay duplicados'

### Analisis del tipo de variable

Analizo el tipo de variable

In [14]:
herramientas.verifica_tipo_y_nulos(df_hecho)

nombre_campo                                         tipo_datos  \
0                      Id                                    [<class 'str'>]   
1    Cantidad de victimas                                    [<class 'int'>]   
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                     Año                                    [<class 'int'>]   
4                     Mes                                    [<class 'int'>]   
5                     Dia                                    [<class 'int'>]   
6                    Hora  [<class 'datetime.time'>, <class 'str'>, <clas...   
7          Hora en entero                     [<class 'int'>, <class 'str'>]   
8         Lugar del hecho                                    [<class 'str'>]   
9           Tipo de calle                                    [<class 'str'>]   
10                  Calle                   [<class 'str'>, <class 'float'>]   
11                  Cruce                                    [<class 'str'>]   
12  Dirección normalizada                                    [<class 'str'>]   
13                 Comuna                                    [<class 'int'>]   
14              XY (CABA)                                    [<class 'str'>]   
15                  Pos x                                    [<class 'str'>]   
16                  Pos y                                    [<class 'str'>]   
17          Participantes                                    [<class 'str'>]   
18                Victima                                    [<class 'str'>]   
19                Acusado                                    [<class 'str'>]   

    no_nulos_%  nulos_%  nulos  
0       100.00     0.00      0  
1       100.00     0.00      0  
2       100.00     0.00      0  
3       100.00     0.00      0  
4       100.00     0.00      0  
5       100.00     0.00      0  
6       100.00     0.00      0  
7       100.00     0.00      0  
8       100.00     0.00      0  
9       100.00     0.00      0  
10       99.86     0.14      1  
11      100.00     0.00      0  
12      100.00     0.00      0  
13      100.00     0.00      0  
14      100.00     0.00      0  
15      100.00     0.00      0  
16      100.00     0.00      0  
17      100.00     0.00      0  
18      100.00     0.00      0  
19      100.00     0.00      0

In [15]:
df_hecho.head(5)

Id  Cantidad de victimas      Fecha   Año  Mes  Dia      Hora  \
0  2016-0001                     1 2016-01-01  2016    1    1  04:00:00   
1  2016-0002                     1 2016-01-02  2016    1    2  01:15:00   
2  2016-0003                     1 2016-01-03  2016    1    3  07:00:00   
3  2016-0004                     1 2016-01-10  2016    1   10  00:00:00   
4  2016-0005                     1 2016-01-21  2016    1   21  05:20:00   

  Hora en entero                            Lugar del hecho Tipo de calle  \
0              4  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA   
1              1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ   
2              7                         AV ENTRE RIOS 2034       AVENIDA   
3              0      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA   
4              5  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA   

              Calle Cruce                              Dirección normalizada  \
0  PIEDRA BUENA AV.    Si  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...   
1    PAZ, GRAL. AV.    Si               PAZ, GRAL. AV. y DE LOS CORRALES AV.   
2    ENTRE RIOS AV.    No                                ENTRE RIOS AV. 2034   
3    LARRAZABAL AV.    Si          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.   
4      SAN JUAN AV.    Si            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.   

   Comuna                                XY (CABA)         Pos x  \
0       8    Point (98896.78238426 93532.43437792)  -58.47533969   
1       9    Point (95832.05571093 95505.41641999)  -58.50877521   
2       1   Point (106684.29090040 99706.57687843)  -58.39040293   
3       8    Point (99840.65224780 94269.16534422)  -58.46503904   
4       1  Point (106980.32827929 100752.16915795)  -58.38718297   

          Pos y   Participantes Victima    Acusado  
0  -34.68757022       MOTO-AUTO    MOTO       AUTO  
1  -34.66977709  AUTO-PASAJEROS    AUTO  PASAJEROS  
2  -34.63189362       MOTO-AUTO    MOTO       AUTO  
3  -34.68092974         MOTO-SD    MOTO         SD  
4  -34.62246630  MOTO-PASAJEROS    MOTO  PASAJEROS

Se observa que los valores de las columnas 'Hora en entero', 'Hora' y en 'Calle' no son solo del tipo entero, por lo tanto se modifica de tal forma que quede solo en valor entero. Inicialmente se identifica la cantidad de valores dependiendo el tipo de dato

In [16]:
# Cantidad de valores por tipo de dato en la columna 'hora'
df_hecho['Hora'].apply(type).value_counts()

Hora
<class 'datetime.time'>        608
<class 'str'>                   85
<class 'datetime.datetime'>      3
Name: count, dtype: int64

In [17]:
df_hecho['Hora en entero'].apply(type).value_counts()

Hora en entero
<class 'int'>    695
<class 'str'>      1
Name: count, dtype: int64

In [18]:
df_hecho['Calle'].apply(type).value_counts()

Calle
<class 'str'>      695
<class 'float'>      1
Name: count, dtype: int64

- **Columna: 'Hora'**

Convierto en datetime todos los valores de la columna 'Hora'

In [19]:
# Se cambia el tipo de dato
df_hecho['Hora'] = df_hecho['Hora'].apply(lambda x: herramientas.convertir_a_time(x))

Verifico la cantidad de valores por tipo de dato en la columna 'hora'

In [20]:
print('Tipos de datos:')
print(df_hecho['Hora'].apply(type).value_counts())
print('Registro con NoneType:')
df_hecho[df_hecho['Hora'].isna()]

Tipos de datos:
Hora
<class 'datetime.time'>    695
<class 'NoneType'>           1
Name: count, dtype: int64
Registro con NoneType:


Id  Cantidad de victimas      Fecha   Año  Mes  Dia  Hora  \
518  2019-0103                     1 2019-12-18  2019   12   18  None   

    Hora en entero          Lugar del hecho Tipo de calle           Calle  \
518             SD  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  PAZ, GRAL. AV.   

    Cruce    Dirección normalizada  Comuna  \
518    Si  PAZ, GRAL. AV. y GRIVEO      11   

                                  XY (CABA)         Pos x         Pos y  \
518  Point (94643.11254058 103831.57115061)  -58.52169422  -34.59471640   

    Participantes Victima Acusado  
518     MOTO-MOTO    MOTO    MOTO

Calculo la hora mas comun y reemplazo el unico valor None , con la misma 

In [21]:
hora_comun = df_hecho['Hora'].mode().iloc[0]
print(f'Hora mas comun: {hora_comun}')
# Se reemplaza el valor None por la hora más común
df_hecho['Hora'].fillna(hora_comun, inplace=True)
# se verifica el tipo de dato para la columna
df_hecho['Hora'].apply(type).value_counts()

Hora mas comun: 09:00:00


Hora
<class 'datetime.time'>    696
Name: count, dtype: int64

- **Columna: 'Hora en entero'**

In [22]:
# Se imputa la hora moda al dato faltante
df_hecho['Hora en entero'] = df_hecho['Hora en entero'].apply(lambda x: int(hora_comun.hour) if x == "SD" else x)
# Se verifica el tipo de dato
df_hecho['Hora en entero'].apply(type).value_counts()

Hora en entero
<class 'int'>    696
Name: count, dtype: int64

In [23]:
# Se verifica el registro completado
df_hecho[df_hecho['Id']=='2019-0103']

Id  Cantidad de victimas      Fecha   Año  Mes  Dia      Hora  \
518  2019-0103                     1 2019-12-18  2019   12   18  09:00:00   

     Hora en entero          Lugar del hecho Tipo de calle           Calle  \
518               9  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  PAZ, GRAL. AV.   

    Cruce    Dirección normalizada  Comuna  \
518    Si  PAZ, GRAL. AV. y GRIVEO      11   

                                  XY (CABA)         Pos x         Pos y  \
518  Point (94643.11254058 103831.57115061)  -58.52169422  -34.59471640   

    Participantes Victima Acusado  
518     MOTO-MOTO    MOTO    MOTO

- **Columna: 'Calle'**

In [24]:
# Se verifica el tipo de dato
print('Tipos de datos:')
print(df_hecho['Hora en entero'].apply(type).value_counts())
# Se observa el registro con tipo de dato float
print('Registro con tipo float:')
df_hecho[df_hecho['Calle'].apply(lambda x: isinstance(x, float))]

Tipos de datos:
Hora en entero
<class 'int'>    696
Name: count, dtype: int64
Registro con tipo float:


Id  Cantidad de victimas      Fecha   Año  Mes  Dia      Hora  \
119  2016-0151                     1 2016-11-18  2016   11   18  20:35:00   

     Hora en entero Lugar del hecho Tipo de calle Calle Cruce  \
119              20              SD         CALLE   NaN    No   

    Dirección normalizada  Comuna    XY (CABA) Pos x Pos y Participantes  \
119                    SD       0  Point (. .)     .     .     PEATON-SD   

    Victima Acusado  
119  PEATON      SD

In [25]:
# Se imputa SD al dato faltante
df_hecho['Calle'].fillna('SD', inplace=True)
# Se verifica el registro completado
df_hecho[df_hecho['Id']=='2016-0151']

Id  Cantidad de victimas      Fecha   Año  Mes  Dia      Hora  \
119  2016-0151                     1 2016-11-18  2016   11   18  20:35:00   

     Hora en entero Lugar del hecho Tipo de calle Calle Cruce  \
119              20              SD         CALLE    SD    No   

    Dirección normalizada  Comuna    XY (CABA) Pos x Pos y Participantes  \
119                    SD       0  Point (. .)     .     .     PEATON-SD   

    Victima Acusado  
119  PEATON      SD

Verificamos si los cambios se realizaron correctamente

In [26]:
herramientas.verifica_tipo_y_nulos(df_hecho)

nombre_campo                                         tipo_datos  \
0                      Id                                    [<class 'str'>]   
1    Cantidad de victimas                                    [<class 'int'>]   
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                     Año                                    [<class 'int'>]   
4                     Mes                                    [<class 'int'>]   
5                     Dia                                    [<class 'int'>]   
6                    Hora                          [<class 'datetime.time'>]   
7          Hora en entero                                    [<class 'int'>]   
8         Lugar del hecho                                    [<class 'str'>]   
9           Tipo de calle                                    [<class 'str'>]   
10                  Calle                                    [<class 'str'>]   
11                  Cruce                                    [<class 'str'>]   
12  Dirección normalizada                                    [<class 'str'>]   
13                 Comuna                                    [<class 'int'>]   
14              XY (CABA)                                    [<class 'str'>]   
15                  Pos x                                    [<class 'str'>]   
16                  Pos y                                    [<class 'str'>]   
17          Participantes                                    [<class 'str'>]   
18                Victima                                    [<class 'str'>]   
19                Acusado                                    [<class 'str'>]   

    no_nulos_%  nulos_%  nulos  
0        100.0      0.0      0  
1        100.0      0.0      0  
2        100.0      0.0      0  
3        100.0      0.0      0  
4        100.0      0.0      0  
5        100.0      0.0      0  
6        100.0      0.0      0  
7        100.0      0.0      0  
8        100.0      0.0      0  
9        100.0      0.0      0  
10       100.0      0.0      0  
11       100.0      0.0      0  
12       100.0      0.0      0  
13       100.0      0.0      0  
14       100.0      0.0      0  
15       100.0      0.0      0  
16       100.0      0.0      0  
17       100.0      0.0      0  
18       100.0      0.0      0  
19       100.0      0.0      0

- **Columnas: 'Acusado' y 'Victima'**

In [27]:
df_hecho['Acusado'].unique()

array(['AUTO', 'PASAJEROS', 'SD', 'OBJETO FIJO', 'CARGAS', 'MOTO',
       'MULTIPLE', 'OTRO', 'BICICLETA', 'TREN'], dtype=object)

In [28]:
df_hecho['Victima'].unique()

array(['MOTO', 'AUTO', 'PEATON', 'SD', 'CARGAS', 'BICICLETA', 'PASAJEROS',
       'MOVIL', 'OBJETO FIJO', 'PEATON_MOTO'], dtype=object)

Para 'Acusado' no se observan características distintas a las esperadas. En cambio, para las categorías de 'Víctima' se observan dos categorías Objeto fijo y Peatón moto que no están en el diccionario de datos. Se observan estos registros para corregirlos.

In [29]:
df_hecho[df_hecho['Victima'].isin(['OBJETO FIJO', 'PEATON_MOTO'])]

Id  Cantidad de victimas      Fecha   Año  Mes  Dia      Hora  \
230  2017-0108                     2 2017-09-02  2017    9    2  04:53:08   
583  2020-0063                     2 2020-12-05  2020   12    5  07:10:00   

     Hora en entero           Lugar del hecho Tipo de calle           Calle  \
230               4   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   
583               7  NUEVA YORK Y ALTA GRACIA         CALLE      NUEVA YORK   

    Cruce     Dirección normalizada  Comuna  \
230    Si  PAZ, GRAL. AV. y MACHAIN      12   
583    Si  NUEVA YORK y ALTA GRACIA      11   

                                  XY (CABA)         Pos x         Pos y  \
230  Point (97098.48468623 109019.96106626)  -58.49491054  -34.54795581   
583  Point (94080.62190808 102083.62453795)  -58.52783814  -34.61047001   

        Participantes      Victima Acusado  
230  AUTO-OBJETO FIJO  OBJETO FIJO    AUTO  
583  PEATON_MOTO-MOTO  PEATON_MOTO    MOTO

Se observan dos registros, uno por cada término distinto al del diccionario de datos. Se decide cambiarlo por OTRO dado que no hay otro dato que pueda indicar el vehículo de la victima.

In [30]:
# Se cambia por OTRO
df_hecho['Victima'] = df_hecho['Victima'].replace({'OBJETO FIJO':'OTRO', 'PEATON_MOTO':'OTRO'})
# Se verifican esos cambios
df_hecho[df_hecho['Id'].isin(['2017-0108', '2020-0063'])]

Id  Cantidad de victimas      Fecha   Año  Mes  Dia      Hora  \
230  2017-0108                     2 2017-09-02  2017    9    2  04:53:08   
583  2020-0063                     2 2020-12-05  2020   12    5  07:10:00   

     Hora en entero           Lugar del hecho Tipo de calle           Calle  \
230               4   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   
583               7  NUEVA YORK Y ALTA GRACIA         CALLE      NUEVA YORK   

    Cruce     Dirección normalizada  Comuna  \
230    Si  PAZ, GRAL. AV. y MACHAIN      12   
583    Si  NUEVA YORK y ALTA GRACIA      11   

                                  XY (CABA)         Pos x         Pos y  \
230  Point (97098.48468623 109019.96106626)  -58.49491054  -34.54795581   
583  Point (94080.62190808 102083.62453795)  -58.52783814  -34.61047001   

        Participantes Victima Acusado  
230  AUTO-OBJETO FIJO    OTRO    AUTO  
583  PEATON_MOTO-MOTO    OTRO    MOTO

- **Columnas: 'XY_(CABA)', 'Pos_x' y 'Pos_y'.**

In [31]:
df_hecho['XY (CABA)']

0        Point (98896.78238426 93532.43437792)
1        Point (95832.05571093 95505.41641999)
2       Point (106684.29090040 99706.57687843)
3        Point (99840.65224780 94269.16534422)
4      Point (106980.32827929 100752.16915795)
                        ...                   
691     Point (102728.60090138 98186.24929177)
692      Point (99624.29795829 97569.69801131)
693     Point (99116.45492358 101045.23284826)
694      Point (99324.54463985 97676.26932409)
695    Point (107664.16647795 104708.63962087)
Name: XY (CABA), Length: 696, dtype: object

Se observan los valores faltantes en las columnas a analaizar 

In [32]:
df_hecho[df_hecho['XY (CABA)']=='Point (. .)']

Id  Cantidad de victimas      Fecha   Año  Mes  Dia      Hora  \
35   2016-0049                     1 2016-04-17  2016    4   17  00:00:00   
38   2016-0052                     1 2016-04-20  2016    4   20  20:00:00   
71   2016-0096                     1 2016-07-25  2016    7   25  07:00:00   
106  2016-0136                     1 2016-10-25  2016   10   25  00:00:00   
119  2016-0151                     1 2016-11-18  2016   11   18  20:35:00   
139  2016-0174                     1 2016-12-27  2016   12   27  00:00:00   
176  2017-0042                     1 2017-04-10  2017    4   10  09:00:00   
180  2017-0050                     2 2017-04-28  2017    4   28  11:08:08   
181  2017-0051                     1 2017-05-01  2017    5    1  03:47:47   
256  2017-0140                     1 2017-11-19  2017   11   19  23:22:17   
313  2018-0039                     1 2018-04-21  2018    4   21  22:15:00   
546  2020-0026                     1 2020-05-17  2020    5   17  06:40:00   
559  2020-0039                     1 2020-09-01  2020    9    1  19:17:42   
621  2021-0023                     1 2021-03-01  2021    3    1  09:20:00   

     Hora en entero                                    Lugar del hecho  \
35                0  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....   
38               20                         AUTOPISTA LUGONES PK 10000   
71                7       AUTOPISTA DELLEPIANE LUIS TTE. GRAL. KM. 2,3   
106               0                   AU BUENOS AIRES - LA PLATA KM. 4   
119              20                                                 SD   
139               0                               AUTOPISTA 25 DE MAYO   
176               9                      AV. LEOPOLDO LUGONES PKM 6900   
180              11            AU PERITO MORENO Y RAMAL ENLACE AU1/AU6   
181               3                                 AU DELLEPIANE 2400   
256              23                        AU ARTURO FRONDIZI PKM 3100   
313              22                           AUTOPISTA LUGONES KM 4.7   
546               6                       LUGONES, LEOPOLDO AV. KM 6,1   
559              19                                    MURGUIONDO 2700   
621               9                   AU BUENOS AIRES LA PLATA  KM 4,5   

    Tipo de calle                                       Calle Cruce  \
35      AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI    No   
38      AUTOPISTA                       LUGONES, LEOPOLDO AV.    No   
71      AUTOPISTA        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.    No   
106     AUTOPISTA           AUTOPISTA BUENOS AIRES - LA PLATA    No   
119         CALLE                                          SD    No   
139     AUTOPISTA                        AUTOPISTA 25 DE MAYO    No   
176      GRAL PAZ                       LUGONES, LEOPOLDO AV.    No   
180     AUTOPISTA                     AUTOPISTA PERITO MORENO    No   
181     AUTOPISTA        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.    No   
256     AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI    No   
313     AUTOPISTA                       LUGONES, LEOPOLDO AV.    No   
546     AUTOPISTA                       LUGONES, LEOPOLDO AV.    No   
559         CALLE                                  MURGUIONDO    No   
621     AUTOPISTA           AUTOPISTA BUENOS AIRES - LA PLATA    No   

                          Dirección normalizada  Comuna    XY (CABA)  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4  Point (. .)   
38                                           SD      13  Point (. .)   
71         AUTOPISTA DELLEPIANE LUIS TTE. GRAL.       8  Point (. .)   
106                                          SD       4  Point (. .)   
119                                          SD       0  Point (. .)   
139                        AUTOPISTA 25 DE MAYO       0  Point (. .)   
176                       LUGONES, LEOPOLDO AV.      14  Point (. .)   
180                                          SD       9  Point (. .)   
181                            

Se observan 14 registros que no presentan las coordenadas planas del hecho y también se ven 12 registros que tampoco presentan los datos 'Pos x' y 'Pos y'. En este caso , dichos valores se reemplazaran por el valor **0** de modo tal que sirva para excluir el valor en el dashboard a realizar.

In [33]:
# Se reemplazan los valores faltantes por 0
df_hecho['Pos x'] = df_hecho['Pos x'].replace('.', 0)
df_hecho['Pos y'] = df_hecho['Pos y'].replace('.', 0)
df_hecho['XY (CABA)'] = df_hecho['XY (CABA)'].replace('Point (. .)', 0)


Se verifica si los cambios anteriores fueron aplicados

In [34]:
df_hecho[df_hecho['Id']=='2016-0174']

Id  Cantidad de victimas      Fecha   Año  Mes  Dia      Hora  \
139  2016-0174                     1 2016-12-27  2016   12   27  00:00:00   

     Hora en entero       Lugar del hecho Tipo de calle                 Calle  \
139               0  AUTOPISTA 25 DE MAYO     AUTOPISTA  AUTOPISTA 25 DE MAYO   

    Cruce Dirección normalizada  Comuna XY (CABA) Pos x Pos y Participantes  \
139    No  AUTOPISTA 25 DE MAYO       0         0     0     0         SD-SD   

    Victima Acusado  
139      SD      SD

### Carga de la Hoja 'HECHOS'

Se procede a guardar los cambios anteriormente realizados en un archivo llamado  **'hechos_limpio.csv'**

In [35]:
df_hecho_limpio = 'data/hechos_limpio.csv'
df_hecho.to_csv(df_hecho_limpio, index=False, encoding='utf-8')
print(f'Se guardó el archivo {df_hecho_limpio}')

Se guardó el archivo data/hechos_limpio.csv
